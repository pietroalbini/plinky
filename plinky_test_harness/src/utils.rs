use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
