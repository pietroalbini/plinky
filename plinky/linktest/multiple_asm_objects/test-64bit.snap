---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  class: Elf64, endian: Little, abi: SystemV, machine: X86_64
 │
 │  section .comment#10 (no perms) in <plinky>
 │   │
 │   │  zero-terminated strings should be deduplicated
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 4c 69 6e 6b 65 72 3a 20 70 6c 69 6e 6b 79 00 │ Linker: plinky. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .rodata#6 (perms: r) in goodbye.o
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 47 6f 6f 64 62 79 65 20 77 6f 72 6c 64 21 0a │ Goodbye world!. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .rodata#16 (perms: r) in hello.o
 │   │
 │   │  ╭────────────────────────────────────────┬───────────────╮
 │   │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a │ Hello world!. │
 │   │  ╰────────────────────────────────────────┴───────────────╯
 │   ┴
 │
 │  section names section .shstrtab#0 in <plinky>
 │   │
 │   │  section names
 │   ┴
 │
 │  string table (for symbols) section .strtab#20 in <plinky>
 │   │
 │   │  symbols view: symbol table
 │   ┴
 │
 │  symbols table section .symtab#21 in <plinky>
 │   │
 │   │  view: symbol table
 │   │  strings: .strtab#20
 │   ┴
 │
 │  section .text#2 (perms: rx) in goodbye.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 00 00 00 ba │ ................ │
 │   │  │ 0f 00 00 00 cd 80 b0 01 bb 00 00 00 00 cd 80    │ ...............  │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭────────────┬─────────────────────┬────────┬────────╮
 │   │  │ Type       │ Symbol              │ Offset │ Addend │
 │   │  ├────────────┼─────────────────────┼────────┼────────┤
 │   │  │ Absolute32 │ <section .rodata#6> │ 0xb    │ 0x0    │
 │   │  ╰────────────┴─────────────────────┴────────┴────────╯
 │   ┴
 │
 │  section .text#12 (perms: rx) in hello.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 00 00 00 ba │ ................ │
 │   │  │ 0d 00 00 00 cd 80 e9 00 00 00 00                │ ...........      │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭────────────┬──────────────────────┬────────┬────────────────────╮
 │   │  │ Type       │ Symbol               │ Offset │ Addend             │
 │   │  ├────────────┼──────────────────────┼────────┼────────────────────┤
 │   │  │ Absolute32 │ <section .rodata#16> │ 0xb    │ 0x0                │
 │   │  ├────────────┼──────────────────────┼────────┼────────────────────┤
 │   │  │ PLT32      │ goodbye#7            │ 0x17   │ 0xfffffffffffffffc │
 │   │  ╰────────────┴──────────────────────┴────────┴────────────────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭──────────────────────┬─────────┬───────────┬────────────┬──────────────────╮
 │  │ Name                 │ Type    │ Source    │ Visibility │ Value            │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ <symbol#0>           │ none    │ <plinky>  │ local      │ <null>           │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ <section .rodata#6>  │ section │ goodbye.o │ local      │ .rodata#6 + 0x0  │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ <section .rodata#16> │ section │ hello.o   │ local      │ .rodata#16 + 0x0 │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ _start#1             │ none    │ hello.o   │ global     │ .text#12 + 0x0   │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ goodbye#7            │ none    │ goodbye.o │ global     │ .text#2 + 0x0    │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ len#6                │ none    │ goodbye.o │ local      │ 0xf              │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ len#12               │ none    │ hello.o   │ local      │ 0xd              │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ msg#5                │ none    │ goodbye.o │ local      │ .rodata#6 + 0x0  │
 │  ├──────────────────────┼─────────┼───────────┼────────────┼──────────────────┤
 │  │ msg#11               │ none    │ hello.o   │ local      │ .rodata#16 + 0x0 │
 │  ╰──────────────────────┴─────────┴───────────┴────────────┴──────────────────╯
 ┴

debug print: calculated layout
 │
 │    Layout:
 │  ╭──────────────────┬─────────────┬─────────────┬────────────────┬───────────────╮
 │  │ Part             │ File offset │ File length │ Memory address │ Memory length │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <elf header>     │ 0x0         │ 0x40        │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <program header> │ 0x40        │ 0xe0        │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <section header> │ 0x120       │ 0x280       │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .shstrtab#0      │ 0x3a0       │ 0x49        │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .comment#10      │ 0x3e9       │ 0xf         │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .strtab#20       │ 0x3f8       │ 0x49        │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .symtab#21       │ 0x441       │ 0xf0        │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>        │ 0x531       │ 0xacf       │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rodata#6        │ 0x1000      │ 0xf         │ 0x400000       │ 0xf           │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rodata#16       │ 0x100f      │ 0xd         │ 0x40000f       │ 0xd           │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>        │ 0x101c      │ 0xfe4       │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#2          │ 0x2000      │ 0x1f        │ 0x401000       │ 0x1f          │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#12         │ 0x201f      │ 0x1b        │ 0x40101f       │ 0x1b          │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>        │ 0x203a      │ 0xfc6       │ -              │ -             │
 │  ├──────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .got.plt#22      │ 0x3000      │ 0x8         │ 0x402000       │ 0x8           │
 │  ╰──────────────────┴─────────────┴─────────────┴────────────────┴───────────────╯
 │
 │    Segments:
 │  ╭──────────┬────────┬───────────┬───────┬─────────────╮
 │  │ Start    │ Align  │ Type      │ Perms │ Content     │
 │  ├──────────┼────────┼───────────┼───────┼─────────────┤
 │  │ 0x400000 │ 0x1000 │ program   │ R     │ .rodata#6   │
 │  │          │        │           │       │ .rodata#16  │
 │  ├──────────┼────────┼───────────┼───────┼─────────────┤
 │  │ 0x401000 │ 0x1000 │ program   │ RX    │ .text#2     │
 │  │          │        │           │       │ .text#12    │
 │  ├──────────┼────────┼───────────┼───────┼─────────────┤
 │  │ 0x402000 │ 0x1000 │ program   │ RW    │ .got.plt#22 │
 │  ├──────────┼────────┼───────────┼───────┼─────────────┤
 │  │ -        │ 0x1    │ GNU stack │ RW    │             │
 │  ╰──────────┴────────┴───────────┴───────┴─────────────╯
 ┴

debug print: object after relocations are applied
 │
 │  class: Elf64, endian: Little, abi: SystemV, machine: X86_64
 │
 │  section .comment#10 (no perms) in <plinky>
 │   │
 │   │  zero-terminated strings should be deduplicated
 │   │
 │   │  not allocated in the resulting memory
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 4c 69 6e 6b 65 72 3a 20 70 6c 69 6e 6b 79 00 │ Linker: plinky. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .got.plt#22 (perms: rw) in <plinky>
 │   │
 │   │  address: 0x402000
 │   │
 │   │  ╭─────────────────────────┬──────────╮
 │   │  │ 00 10 40 00 00 00 00 00 │ ..@..... │
 │   │  ╰─────────────────────────┴──────────╯
 │   ┴
 │
 │  section .rodata#6 (perms: r) in goodbye.o
 │   │
 │   │  address: 0x400000
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 47 6f 6f 64 62 79 65 20 77 6f 72 6c 64 21 0a │ Goodbye world!. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .rodata#16 (perms: r) in hello.o
 │   │
 │   │  address: 0x40000f
 │   │
 │   │  ╭────────────────────────────────────────┬───────────────╮
 │   │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a │ Hello world!. │
 │   │  ╰────────────────────────────────────────┴───────────────╯
 │   ┴
 │
 │  section names section .shstrtab#0 in <plinky>
 │   │
 │   │  section names
 │   ┴
 │
 │  string table (for symbols) section .strtab#20 in <plinky>
 │   │
 │   │  symbols view: symbol table
 │   ┴
 │
 │  symbols table section .symtab#21 in <plinky>
 │   │
 │   │  view: symbol table
 │   │  strings: .strtab#20
 │   ┴
 │
 │  section .text#2 (perms: rx) in goodbye.o
 │   │
 │   │  address: 0x401000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 00 40 00 ba │ .............@.. │
 │   │  │ 0f 00 00 00 cd 80 b0 01 bb 00 00 00 00 cd 80    │ ...............  │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#12 (perms: rx) in hello.o
 │   │
 │   │  address: 0x40101f
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 0f 00 40 00 ba │ .............@.. │
 │   │  │ 0d 00 00 00 cd 80 e9 c6 ff ff ff                │ ...........      │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭──────────────────────────┬─────────┬───────────┬────────────┬───────────────────╮
 │  │ Name                     │ Type    │ Source    │ Visibility │ Value             │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ <symbol#0>               │ none    │ <plinky>  │ local      │ <null>            │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ <section .rodata#6>      │ section │ goodbye.o │ local      │ .rodata#6 + 0x0   │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ <section .rodata#16>     │ section │ hello.o   │ local      │ .rodata#16 + 0x0  │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ _GLOBAL_OFFSET_TABLE_#15 │ none    │ <plinky>  │ local      │ .got.plt#22 + 0x0 │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ _start#1                 │ none    │ hello.o   │ global     │ .text#12 + 0x0    │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ goodbye#7                │ none    │ goodbye.o │ global     │ .text#2 + 0x0     │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ len#6                    │ none    │ goodbye.o │ local      │ 0xf               │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ len#12                   │ none    │ hello.o   │ local      │ 0xd               │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ msg#5                    │ none    │ goodbye.o │ local      │ .rodata#6 + 0x0   │
 │  ├──────────────────────────┼─────────┼───────────┼────────────┼───────────────────┤
 │  │ msg#11                   │ none    │ hello.o   │ local      │ .rodata#16 + 0x0  │
 │  ╰──────────────────────────┴─────────┴───────────┴────────────┴───────────────────╯
 ┴

debug print: built elf
 │
 │    Metadata:
 │  ╭────────────┬────────────╮
 │  │ Class      │ ELF 64bit  │
 │  ├────────────┼────────────┤
 │  │ Endian     │ Little     │
 │  ├────────────┼────────────┤
 │  │ ABI        │ System V   │
 │  ├────────────┼────────────┤
 │  │ Machine    │ x86-64     │
 │  ├────────────┼────────────┤
 │  │ Type       │ Executable │
 │  ├────────────┼────────────┤
 │  │ Entrypoint │ 0x40101f   │
 │  ╰────────────┴────────────╯
 │
 │  section #0 (address: 0x0)
 │   │
 │   │  empty section
 │   ┴
 │
 │  section .shstrtab#1 (address: 0x0)
 │   │
 │   │    Strings table:
 │   │  ╭──────┬───────────╮
 │   │  │ 0x0  │           │
 │   │  ├──────┼───────────┤
 │   │  │ 0x1  │ .shstrtab │
 │   │  ├──────┼───────────┤
 │   │  │ 0xb  │ .text     │
 │   │  ├──────┼───────────┤
 │   │  │ 0x11 │ .rodata   │
 │   │  ├──────┼───────────┤
 │   │  │ 0x19 │ .comment  │
 │   │  ├──────┼───────────┤
 │   │  │ 0x22 │ .text     │
 │   │  ├──────┼───────────┤
 │   │  │ 0x28 │ .rodata   │
 │   │  ├──────┼───────────┤
 │   │  │ 0x30 │ .strtab   │
 │   │  ├──────┼───────────┤
 │   │  │ 0x38 │ .symtab   │
 │   │  ├──────┼───────────┤
 │   │  │ 0x40 │ .got.plt  │
 │   │  ╰──────┴───────────╯
 │   ┴
 │
 │  section .text#2 (address: 0x401000)
 │   │
 │   │  program data | permissions: R X
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 00 40 00 ba │ .............@.. │
 │   │  │ 0f 00 00 00 cd 80 b0 01 bb 00 00 00 00 cd 80    │ ...............  │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .rodata#3 (address: 0x400000)
 │   │
 │   │  program data | permissions: R
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 47 6f 6f 64 62 79 65 20 77 6f 72 6c 64 21 0a │ Goodbye world!. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .comment#4 (address: 0x0)
 │   │
 │   │  program data | permissions: -   | deduplicating zero-terminated strings
 │   │
 │   │  ╭──────────────────────────────────────────────┬─────────────────╮
 │   │  │ 4c 69 6e 6b 65 72 3a 20 70 6c 69 6e 6b 79 00 │ Linker: plinky. │
 │   │  ╰──────────────────────────────────────────────┴─────────────────╯
 │   ┴
 │
 │  section .text#5 (address: 0x40101f)
 │   │
 │   │  program data | permissions: R X
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ b8 04 00 00 00 bb 01 00 00 00 b9 0f 00 40 00 ba │ .............@.. │
 │   │  │ 0d 00 00 00 cd 80 e9 c6 ff ff ff                │ ...........      │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .rodata#6 (address: 0x40000f)
 │   │
 │   │  program data | permissions: R
 │   │
 │   │  ╭────────────────────────────────────────┬───────────────╮
 │   │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a │ Hello world!. │
 │   │  ╰────────────────────────────────────────┴───────────────╯
 │   ┴
 │
 │  section .strtab#7 (address: 0x0)
 │   │
 │   │    Strings table:
 │   │  ╭──────┬───────────────────────╮
 │   │  │ 0x0  │                       │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x1  │                       │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x2  │ _GLOBAL_OFFSET_TABLE_ │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x18 │ goodbye.S             │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x22 │ msg                   │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x26 │ len                   │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x2a │ hello.S               │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x32 │ msg                   │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x36 │ len                   │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x3a │ _start                │
 │   │  ├──────┼───────────────────────┤
 │   │  │ 0x41 │ goodbye               │
 │   │  ╰──────┴───────────────────────╯
 │   ┴
 │
 │  section .symtab#8 (address: 0x0)
 │   │
 │   │    Symbol table:
 │   │  ╭─────────────────────────┬─────────┬──────┬────────────┬────────────┬──────────┬──────╮
 │   │  │ Name                    │ Binding │ Type │ Visibility │ Definition │ Value    │ Size │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ #0                      │ Local   │ -    │ Default    │ Undefined  │ 0x0      │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ _GLOBAL_OFFSET_TABLE_#1 │ Local   │ -    │ Default    │ .got.plt#9 │ 0x402000 │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ goodbye.S#2             │ Local   │ File │ Default    │ Absolute   │ 0x0      │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ msg#3                   │ Local   │ -    │ Default    │ .rodata#3  │ 0x400000 │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ len#4                   │ Local   │ -    │ Default    │ Absolute   │ 0xf      │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ hello.S#5               │ Local   │ File │ Default    │ Absolute   │ 0x0      │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ msg#6                   │ Local   │ -    │ Default    │ .rodata#6  │ 0x40000f │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ len#7                   │ Local   │ -    │ Default    │ Absolute   │ 0xd      │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ _start#8                │ Global  │ -    │ Default    │ .text#5    │ 0x40101f │ 0x0  │
 │   │  ├─────────────────────────┼─────────┼──────┼────────────┼────────────┼──────────┼──────┤
 │   │  │ goodbye#9               │ Global  │ -    │ Default    │ .text#2    │ 0x401000 │ 0x0  │
 │   │  ╰─────────────────────────┴─────────┴──────┴────────────┴────────────┴──────────┴──────╯
 │   ┴
 │
 │  section .got.plt#9 (address: 0x402000)
 │   │
 │   │  program data | permissions: RW
 │   │
 │   │  ╭─────────────────────────┬──────────╮
 │   │  │ 00 10 40 00 00 00 00 00 │ ..@..... │
 │   │  ╰─────────────────────────┴──────────╯
 │   ┴
 │
 │    Segments:
 │  ╭───────────┬───────┬──────────┬─────────────┬──────────┬────────────────┬────────────╮
 │  │ Type      │ Perms │ Aligment │ File offset │ File len │ Memory address │ Memory len │
 │  ├───────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load      │ R     │ 0x1000   │ 0x1000      │ 0x1c     │ 0x400000       │ 0x1c       │
 │  ├───────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load      │ R X   │ 0x1000   │ 0x2000      │ 0x3a     │ 0x401000       │ 0x3a       │
 │  ├───────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load      │ RW    │ 0x1000   │ 0x3000      │ 0x8      │ 0x402000       │ 0x8        │
 │  ├───────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ GNU stack │ RW    │ 0x1      │ 0x0         │ 0x0      │ 0x0            │ 0x0        │
 │  ╰───────────┴───────┴──────────┴─────────────┴──────────┴────────────────┴────────────╯
 ┴

==============

running exited with exit status: 0

=== stdout ===
Hello world!
Goodbye world!

no stderr present
